File: adt2pack/i386/LZSS_decompress.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure GetChar; assembler;
asm
        push    ebx
        mov     bx,ibufCount
        cmp     bx,ibufSize
        jb      @@1
        jmp     @@2
@@1:    push    edi
        mov     edi,input_ptr
        mov     al,byte ptr [edi+ebx]
        pop     edi
        inc     ebx
        mov     ibufCount,bx
        pop     ebx
        clc
        jmp     @@3
@@2:    pop     ebx
        stc
@@3:
end;

procedure PutChar; assembler;
asm
        push    ebx
        mov     bx,output_size
        push    edi
        mov     edi,output_ptr
        mov     byte ptr [edi+ebx],al
        pop     edi
        inc     ebx
        mov     output_size,bx
        pop     ebx
end;

procedure LZSS_decode;
begin
  asm
        mov     ibufCount,0
        mov     ax,input_size
        mov     ibufSize,ax
        mov     output_size,0
        xor     ebx,ebx
        xor     edx,edx
        mov     edi,N-F
@@1:    shr     dx,1
        or      dh,dh
        jnz     @@2
        call    GetChar
        jc      @@5
        mov     dh,0ffh
        mov     dl,al
@@2:    test    dx,1
        jz      @@3
        call    GetChar
        jc      @@5
        push    esi
        mov     esi,work_ptr
        add     esi,edi
        mov     byte ptr [esi],al
        pop     esi
        inc     edi
        and     edi,N-1
        call    PutChar
        jmp     @@1
@@3:    call    GetChar
        jc      @@5
        mov     ch,al
        call    GetChar
        jc      @@5
        mov     bh,al
        mov     cl,4
        shr     bh,cl
        mov     bl,ch
        mov     cl,al
        and     cl,0fh
        add     cl,THRESHOLD
        inc     cl
@@4:    and     ebx,N-1
        push    esi
        mov     esi,work_ptr
        mov     al,byte ptr [esi+ebx]
        add     esi,edi
        mov     byte ptr [esi],al
        pop     esi
        inc     edi
        and     edi,N-1
        call    PutChar
        inc     ebx
        dec     cl
        jnz     @@4
        jmp     @@1
@@5:
  end;
end;

File: adt2pack/LZSS_decompress.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

#define N_BITS 12
#define F_BITS 4
#define THRESHOLD 2
#define N (1 << N_BITS)
#define F ((1 << F_BITS) + THRESHOLD)

typedef struct
{
  const uint8_t *input_ptr;
  uint16_t input_pos;
  uint16_t input_size;
  uint8_t *output_ptr;
  uint16_t output_size;
  uint8_t last_ch;
  uint8_t dict[0x10000]; // 64KiB
} LZSSDecoderState_t;

static bool LZSS_GetChar (LZSSDecoderState_t *ds)
{
  if (ds->input_pos < ds->input_size)
  {
    ds->last_ch = ds->input_ptr[ds->input_pos++];
    return false;
  }
  else
    return true;
}

static void LZSS_PutChar (LZSSDecoderState_t *ds)
{
  ds->output_ptr[ds->output_size++] = ds->last_ch;
}

static void LZSS_decode (LZSSDecoderState_t *ds)
{
  uint16_t d = 0;
  uint32_t j = N - F;

  ds->input_pos = 0;
  ds->output_size = 0;

  for (;;)
  {
    d >>= 1;
    if ((d & 0xFF00) == 0)
    {
      if (LZSS_GetChar (ds)) break;
      d = 0xFF00 + ds->last_ch;
    }
    if ((d & 1) != 0)
    {
      if (LZSS_GetChar (ds)) break;
      ds->dict[j++] = ds->last_ch;
      j &= N - 1;
      LZSS_PutChar (ds);
    }
    else
    {
      uint32_t i;
      uint8_t a, b;

      if (LZSS_GetChar (ds)) break;
      b = ds->last_ch;
      if (LZSS_GetChar (ds)) break;
      i = ((ds->last_ch << 4) & 0xFF00) + b;
      a = (ds->last_ch & 0x0F) + THRESHOLD + 1;
      do
      {
        i &= N - 1;
        ds->last_ch = ds->dict[i++];
        ds->dict[j++] = ds->last_ch;
        j &= N - 1;
        LZSS_PutChar (ds);
      } while (--a);
    }
  }
}

uint16_t LZSS_decompress (const void *source, void *dest, uint16_t size)
{
  uint_least16_t result = 0;
  LZSSDecoderState_t *ds = NULL;

  if ((ds = malloc (sizeof (LZSSDecoderState_t))) == NULL) goto _exit;

  ds->input_ptr = source;
  //ds->input_pos
  ds->input_size = size;
  ds->output_ptr = dest;
  ds->output_size = 0;
  //ds->last_ch
  memset (ds->dict, 0, sizeof (ds->dict));
  LZSS_decode (ds);
  result = ds->output_size;

_exit:
  if (ds != NULL) free (ds);
  return result;
}
